Refactor TF modules to simplify user experience

eyalzek opened the pull request:
This should not be merged until the [template](https://github.com/doitintl/cmp-main/blob/dev/services/aws-k8s-metrics/delivery/web/handlers/deployments/templates/cluster.tf.tpl#L14) is also updated to reflect the changes.

This repo now contains 2 modules:
1. `region-base` - to be deployed once per region and account (creating S3 bucket and necessary IAM resources).
2. `cluster` - to be deployed once per cluster (creating IAM for EKS Lens to  access the S3 bucket, deploys EKS Lens and registers/deregisters the cluster to the DoiT console).

eyalzek committed:
Remove extra newlines

eyalzek committed:
Restructure TF modules

Create 2 modules:
1. A global module that creates resources which are necessary once
   per account
2. A module for all resources that are necessary per cluster (can be
   created several times per account)

This structure should make it possible to refer to the modules by git
url rather than having to clone them locally.

eyalzek committed:
Create module for regional base resources

This module should be used once per region per account.

Example usage:

```hcl
module "eu-central-1-base" {
  source = "git::https://github.com/doitintl/terraform-eks-lens.git//region-base?ref=feat/refactor"
}
```

eyalzek committed:
Create `cluster` module

For creating all resources necessary for running EKS Lens in a cluster
and deploy it.

This should be used once per cluster:

```hcl
module "<REGION>-<CLUSTER_NAME>" {
  source = "git::https://github.com/doitintl/terraform-eks-lens.git//cluster"

  cluster = {
    cluster_name     = "<CLUSTER_NAME>"
    deployment_id    = "<DEPLOYMENT_ID>"
    kube_state_image = "registry.k8s.io/kube-state-metrics/kube-state-metrics:v2.9.2" # make sure to use the latest available image
    otel_image       = "otel/opentelemetry-collector-contrib:0.83.0" # make sure to use the latest available image
  }
  # If running in EKS:
  cluster_oidc_issuer_url = "<CLUSTER_OIDC_ISSUER_URL>"
  # Alternatively, if managing your own cluster on EC2, set `cluster_oidc_issuer_url` to an empty string and uncomment the following:
  #ec2_cluster = true

  # when configuring multiple providers for different clusters, you can configure the module to use to correct provider alias:
  providers = {
    kubernetes = kubernetes.<PROVIDER_ALIAS>
  }
}
```

eyalzek committed:
Add onboard/offboarding hook

eyalzek committed:
Update root README

and add per-module README files.

eyalzek committed:
Add example providers file

lauraseidler commented on cluster/main.tf:
```diff
@@ -0,0 +1,112 @@
+  s3_policy_arn      = "arn:aws:iam::${local.account_id}:policy/doit_eks_lens_s3_${local.region}"
+  role_arn           = "arn:aws:iam::${local.account_id}:role/doitintl_eks_import_${local.region}"
+  cluster_issuer_url = replace(var.cluster_oidc_issuer_url, "https://", "")
+  doit_webhook_url   = "https://dev-app.doit.com/webhooks/v1/eks-metrics"
```
```suggestion
  doit_webhook_url   = "https://console.doit.com/webhooks/v1/eks-metrics"
```

lauraseidler commented on cluster/main.tf:
```diff
@@ -0,0 +1,112 @@
+  }
+
+  provisioner "local-exec" {
+    when       = destroy
```
I don't think this will work as expected, since if the cluster module is removed entirely, this resource won't be in the configuration anymore, so this won't run: https://developer.hashicorp.com/terraform/language/resources/provisioners/syntax#destroy-time-provisioners

I'm not sure how to work around this best - I was thinking we might be able to leverage a conditional [`prevent_destroy`](https://developer.hashicorp.com/terraform/tutorials/state/resource-lifecycle#prevent-resource-deletion) somehow, but not sure if it's feasible.

eyalzek replied:
Hmm, well I'm assuming that if somebody decides to get rid of this TF configuration they would also destroy it beforehand?

lauraseidler replied:
Hm, would they? I'd expect the just remove the `module` block and apply? At which point, if I understand this correctly, this hook wouldn't run.

eyalzek replied:
This is possible, in which case the README gives the command to offboard... I'm not sure there's a better way to automate this.
Honestly, I would put a TTL on our side to offboard any cluster that is unresponsive for more than X days

lauraseidler replied:
@hobord for your consideration

lauraseidler commented on README.md:
```diff
@@ -68,11 +31,55 @@ terraform plan
+  # Alternatively, if managing your own cluster on EC2, set `cluster_oidc_issuer_url` to an empty string and uncomment the following:
+  #ec2_cluster = true
+
+  # when configuring multiple providers for different clusters, you can configure the module to use to correct provider alias:
+  providers = {
+    kubernetes = kubernetes.<PROVIDER_ALIAS>
+  }
```
Isn't this _required_ as soon as they have more than one cluster per region? Since there will need to be one provider per cluster.

eyalzek replied:
Technically, but you could have a single cluster module and a single kubernetes provider which will be used by default.

This configuration is generated by us anyway, so I suppose it would contain the provider name already? The readme is more of a reference.

I'm open to suggestions if you have a better idea on how to convey this.

lauraseidler commented on cluster/main.tf:
```diff
@@ -0,0 +1,112 @@
+      \"account_id\": \"${local.account_id}\",
+      \"region\": \"${local.region}\",
+      \"cluster_name\": \"${var.cluster.name}\",
+      \"deployment_id\": \"${var.cluster.name}\"
```
```suggestion
      \"deployment_id\": \"${var.cluster.deployment_id}\"
```

eyalzek replied:
good catch

lauraseidler commented on cluster/main.tf:
```diff
@@ -0,0 +1,112 @@
+      \"account_id\": \"${local.account_id}\",
+      \"region\": \"${local.region}\",
+      \"cluster_name\": \"${var.cluster.name}\",
+      \"deployment_id\": \"${var.cluster.name}\"
```
```suggestion
      \"deployment_id\": \"${var.cluster.deployment_id}\"
```

eyalzek committed:
Add option to not deploy manifests

Useful when planning to manage the manifests with another tool (ArgoCD,
Helm, etc).
